commands/verify/verify.go

```diff
@@ -185,64 +243,83 @@ func DoVerifyAccordingConfig() error {
 			retryCount,
 			interval,
 			failFast,
+			summary,
+			writeLock,
 		}
-		stopChan := make(chan bool)
-		goroutineNum := len(e2eConfig.Verify.Cases)
-		waitGroup.Add(goroutineNum)
 
-		for idx, v := range e2eConfig.Verify.Cases {
-			go concurrentVerifySingleCase(idx, v, &ConcurrentErrors, VerifyInfo, &waitGroup, stopChan)
+		waitGroup.Add(caseNumber)
+
+		for idx := range e2eConfig.Verify.Cases {
+			go concurrentVerifySingleCase(idx, &e2eConfig.Verify.Cases[idx], VerifyInfo, &waitGroup, stopChan)
 		}
 
 		if failFast {
-			if shouldExit := check(stopChan, goroutineNum); shouldExit {
-				return errs.ErrorOrNil()
+			if shouldExit := check(stopChan, caseNumber); shouldExit {
+				return nil
 			}
 		}
-
 		waitGroup.Wait()
 	} else {
 		for idx, v := range e2eConfig.Verify.Cases {
+			spinnerLiveText, _ := pterm.DefaultSpinner.WithShowTimer(false).Start()
+			pterm.Error.Prefix = pterm.Prefix{
+				Text:  "DETAILS",
+				Style: &pterm.ThemeDefault.ErrorPrefixStyle,
+			}
 			if v.GetExpected() == "" {
-				errMsg := fmt.Sprintf("the expected data file for case[%v] is not specified\n", idx)
+				errMsg := fmt.Sprintf("failed to verify %v", caseName(v.Name, idx))
+				spinnerLiveText.Warning(errMsg)
+				spinnerLiveText.Fail(fmt.Sprintf("the expected data file for %v is not specified\n", caseName(v.Name, idx)))
+				summary.errNum++
 				if failFast {
-					return errors.New(errMsg)
+					Summary(summary, caseNumber)
+					return nil
 				}
-				logger.Log.Warnf(errMsg)
-				errs = multierror.Append(errs, errors.New(errMsg))
 				continue
 			}
-
-			for current := 1; current <= retryCount; current++ {
+			for current := 0; current <= retryCount; current++ {
 				if err := verifySingleCase(v.GetExpected(), v.GetActual(), v.Query); err == nil {
+					summary.successNum++
+					caseSuccessByCurrent(spinnerLiveText, current, caseName(v.Name, idx), idx)
 					break
 				} else if current != retryCount {
-					logger.Log.Warnf("verify case failure, will continue retry, %v", err)
-					time.Sleep(interval)
+					if current == 0 {
+						spinnerLiveText.UpdateText(fmt.Sprintf("failed to verify %v, will continue retry:", caseName(v.Name, idx)))
+						time.Sleep(time.Second * 2)
+					} else {
+						spinnerLiveText.UpdateText(fmt.Sprintf("failed to verify %v, retry [%d/%d]", caseName(v.Name, idx), current, retryCount))
+						time.Sleep(interval)
+					}
 				} else {
+					summary.errNum++
+					spinnerLiveText.UpdateText(fmt.Sprintf("failed to verify %v, retry [%d/%d]", caseName(v.Name, idx), current, retryCount))
+					time.Sleep(time.Second)
+					spinnerLiveText.Warning(fmt.Sprintf("failed to verify %v, retried %d time(s):", caseName(v.Name, idx), current))
+					spinnerLiveText.Fail(err)
+					fmt.Println()
 					if failFast {
-						return err
+						Summary(summary, caseNumber)
+						return nil
 					}
-					errs = multierror.Append(errs, err)
 				}
 			}
 		}
 	}
-
-	return errs.ErrorOrNil()
+	Summary(summary, caseNumber)
+	return nil
```

Comment:
If there is an error, we still need return it, so infra-e2e can fail CI when it's used as GitHub Action.